File: hackerRank/crackingTheCodingInterview/TriesContacts.java
import java.util.Scanner;

public class TriesContacts{

  public static void main(String args[]){
    Scanner in = new Scanner(System.in);
    Trie t = new Trie();
    int n = in.nextInt();
    for(int a0 = 0; a0 < n; a0++){
        String op = in.next();
        String contact = in.next();
        if(op.equals("add")){
          addWordToTrie(t, contact);
        }
        if(op.equals("find")){
          System.out.println(findPartials(t, contact));
        }
    }
  }

  public static void addWordToTrie(Trie t, String word){
    Trie lastTrie = t;
    for(int i=0; i<word.length(); i++){
      lastTrie = lastTrie.addChild(word.charAt(i));
    }
  }

  public static int findPartials(Trie t, String partial){
    Trie fromTrie = t;
    for(int i=0; i<partial.length(); i++){
      Trie child = fromTrie.getChild(partial.charAt(i));
      if (child != null){
        fromTrie = child;
      }
      else{
        return 0;
      }
    }
    return fromTrie.wordCount;
  }
  
}

class Trie{
  int wordCount;
  Trie[] letters = new Trie[26];

  public static int getIndexFromChar(char c){
    return (int)c - (int)'a';
  }

  public Trie(){
    wordCount = 0;
    for(int i=0; i<26; i++){
      letters[i] = null;
    }
  }

  public Trie addChild(char letter){
    Trie curr = null;
    int index = getIndexFromChar(letter);
    if(letters[index] != null){
      curr = letters[index];
      curr.wordCount++;
    }
    else{
      letters[index] = curr = new Trie();
      curr.wordCount = 1;
    }

    return curr;
  }

  public Trie getChild(char letter){
    return letters[getIndexFromChar(letter)];
  }
}

File: hackerRank/crackingTheCodingInterview/tries_contacts.py

# Solves all but one case, something related with memory allocaded for python
class Trie:
  def __init__(self, is_word=False):
    self.word_count = 0
    self.children = {}

  def add_child(self, letter):
    curr_val = None
    if letter in self.children:
      curr_val = self.children[letter]
      curr_val.word_count += 1
    else:
      self.children[letter] = curr_val = Trie();
      curr_val.word_count = 1
    return curr_val

  def get_child(self, letter):
    if letter in self.children:
      return self.children[letter]

  def __str__(self):
    return str(self.is_word) + str(self.children)

def add_word_to_trie(trie, word):
  last_letter = trie
  for letter in word:
    last_letter = last_letter.add_child(letter)
  last_letter.is_word = True

def find_partial(trie, part):
  from_trie = trie
  for letter in part:
    child = from_trie.get_child(letter)
    if child:
      from_trie = from_trie.children[letter]
    else:
      return 0
  return from_trie.word_count

n = int(input())
tr = Trie()
for x in range(n):
  inp = input().split()
  operation = inp[0]
  parameter = inp[1]

  if operation == 'add':
    add_word_to_trie(tr, parameter)
  if operation == 'find':
    print(find_partial(tr, parameter))